src/map.rs

```diff
@@ -1785,6 +1694,59 @@ impl<K, V> OrderMapCore<K, V> {
     }
 }
 
+trait ProbeAction<'a, Sz: Size, K, V>: Sized {
+    type Output;
+    fn hit(self, entry: OccupiedEntry<'a, K, V>) -> Self::Output;
+    fn empty(self, entry: VacantEntry<'a, K, V>) -> Self::Output;
+    fn steal(self, entry: VacantEntry<'a, K, V>) -> Self::Output {
+        self.empty(entry)
+    }
+}
+
+struct InsertValue<V>(V);
+
+impl<'a, Sz: Size, K, V> ProbeAction<'a, Sz, K, V> for InsertValue<V> {
+    type Output = (usize, Option<V>);
+    fn hit(self, entry: OccupiedEntry<'a, K, V>) -> Self::Output {
+        let old = replace(&mut entry.map.entries[entry.index].value, self.0);
+        (entry.index, Some(old))
+    }
+    fn empty(self, entry: VacantEntry<'a, K, V>) -> Self::Output {
+        let pos = &mut entry.map.indices[entry.probe];
+        let index = entry.map.entries.len();
+        *pos = Pos::with_hash::<Sz>(index, entry.hash);
+        entry.map.entries.push(Bucket {
+            hash: entry.hash,
+            key: entry.key,
+            value: self.0,
+        });
+        (index, None)
+    }
+    fn steal(self, entry: VacantEntry<'a, K, V>) -> Self::Output {
+        let index = entry.map.entries.len();
```

Comment:
this could be just `entry.insert_impl::<Sz>(self.0)` right? Something like that to reduce duplication

Reply:
Ah good idea. Do you mean `empty` could use that? I was afraid that entering the probe loop another time would be costly (technically, you don't have to in the empty case), but it seems to be fine.

Reply:
Nvm, turns out doing it in `empty` isn't great, it basically remove the fast path that the old insert code had, making steal and empty the same.

Reply:
Are we then maybe identifying the main thing that slows down entry compared with the old insert? Maybe if we updated entry's VacantEntry to skip that loop, for the non-steal case, that insert and entry would match each other in performance?